### PR TITLE
ci(linting): remove (nearly) all linting warnings

### PR DESCRIPTION
**Pull request by Leo310**
still need to resolve dependency-cycle warnings.

### CODE REVIEW

**Comment by Leo310** [comment]
The remaining import/no-cycle warnings would require quite some restructuring.
@Jonasdoubleyou @dhenkel92 should I tackle them or do you think we should ignore them for now?

File: .eslintrc
Comment (dhenkel92) [inline .eslintrc:1]: Could you try to add the recomented rules and see what happens? 
I think we are missing some important rules, like `prefer-const`.

```
extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
```

https://typescript-eslint.io/getting-started/#step-2-configuration

File: package.json
Comment (dhenkel92) [inline package.json:1]: We should also think about upgrading the eslint related dependencies.

**Review by dhenkel92 — COMMENTED** [review]
> The remaining import/no-cycle warnings would require quite some restructuring. @Jonasdoubleyou @dhenkel92 should I tackle them or do you think we should ignore them for now?

Maybe we should add the "/* eslint-disable import/no-cycle */" comment to these files as well. This would remove the warnings, but makes sure that we don't introduce new cyclic dependencies.

File: package.json [hunk 1]
```diff
@@ -43,6 +43,7 @@
     "@aws-sdk/client-s3": "^3.369.0",
     "@aws-sdk/s3-request-presigner": "^3.137.0",
     "@prisma/client": "4.15.0",
+    "@types/qrcode": "^1.5.1",
```
[the remark this answers is not in the document]
Comment (Leo310) [reply package.json:46]: my linter was complaining that the declaration file was missing. so installed the types of the qrcode package

File: .eslintrc
Comment (Leo310) [reply .eslintrc:1]: After running `eslint --fix` there still are 968 problems haha.  There are a lot of "no-unused-vars" warnings caused mainly by this:
image
and this:
image

File: package.json
Comment (Leo310) [reply package.json:1]: what do you mean by upgrading it? Most of the packages dont have a fixed version:
image

File: .eslintrc
Comment (dhenkel92) [reply .eslintrc:1]: 😂
What if we remove the no unused car rule?

File: package.json
Comment (dhenkel92) [reply package.json:1]: But it doesn't apply for the major version

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhenkel92) [reply package.json:46]: Ahh nice 😄

File: .eslintrc
Comment (Jonasdoubleyou) [reply .eslintrc:1]: Hm we could remove the "unused function parameter" but I think it actually increases readability here (especially @FieldResolver(returns => Type))

Comment (Jonasdoubleyou) [reply .eslintrc:1]: But I would generally try to first get to 0 warnings, then enforce no linter warnings in CI. Then we can still decide to turn on further rules

Comment (dhenkel92) [reply .eslintrc:1]: What is the advantage of this?

File: package.json
Comment (Leo310) [reply package.json:1]: Ah true

File: .eslintrc
Comment (Leo310) [reply .eslintrc:1]: i turned the no-unused-var rule off and now there are only 156 problems left.
Most of them are unexpected-any's ("no-explicit-any") errors

Comment (dhenkel92) [reply .eslintrc:1]: What if you remove this one as well?
We could create a follow up ticket to fix it as it's only metadata.

**Comment by Leo310** [comment]
Any idea why the integration tests fail now?

**Comment by Leo310** [comment]
Only 3 more errors to resolve 🎉

File: graphql/me/mutation.ts [hunk 2]
```diff
@@ -196,11 +184,7 @@ export class MutateMeResolver {
     @Mutation((returns) => Student)
     @Authorized(Role.UNAUTHENTICATED, Role.ADMIN)
     @RateLimit('RegisterStudent', 10 /* requests per */, 5 * 60 * 60 * 1000 /* 5 hours */)
-    async meRegisterStudent(
-        @Ctx() context: GraphQLContext,
-        @Arg('data') data: RegisterStudentInput,
-        @Arg('noEmail', { nullable: true }) noEmail: boolean = false
-    ) {
+    async meRegisterStudent(@Ctx() context: GraphQLContext, @Arg('data') data: RegisterStudentInput, @Arg('noEmail', { nullable: true }) noEmail = false) {
```
Comment (Jonasdoubleyou) [inline graphql/me/mutation.ts:187]: The :boolean is needed here for TypeGraphQLs reflection (unlile Typescript itself which infers the type from the default parameter) ...

Comment (dhenkel92) [reply graphql/me/mutation.ts:187]: Then let's suppress this one as well 🙂

**Comment by Leo310** [comment]
@Jonasdoubleyou not sure why the integration test failed at the pupilCreatesChat mutation. any idea? Maybe has something to do with the merge

File: common/chat/helper.ts [hunk 3]
```diff
@@ -166,7 +168,7 @@ const convertConversationInfosToString = (conversationInfos: ConversationInfos):
     };
 
     for (const key in conversationInfos.custom) {
-        if (conversationInfos.custom.hasOwnProperty(key)) {
+        if (Object.prototype.hasOwnProperty.call(key)) {
```
Comment (Jonasdoubleyou) [inline common/chat/helper.ts:171]: WTF, `.call(` is missing the first argument, should be `.call(conversationInfos, key)` ...

Comment (Leo310) [reply common/chat/helper.ts:171]: ahh lost, overlooked this when fixing all the linting stuff. mb

Comment (Leo310) [reply common/chat/helper.ts]: didnt seem to fix it :/

**Review by Jonasdoubleyou — DISMISSED** [review]
looks good, haven't yet checked the changes to package.json. Was it just reordered?

**Comment by Leo310** [comment]
> looks good, haven't yet checked the changes to package.json. Was it just reordered?

The only thing that I changed in the package.json was increasing the eslint package version numbers and including the qrcode/types because they were missing. Not sure what all the other changes are

**Review by dhenkel92 — COMMENTED** [review]
Amazing work! 😍 
It was much more effort than expected, but I'm confident that the linter & tsc will help us to find problems before they go to production.

Nevertheless, I'm also not sure what happened to the package.json. From the diff it seems like some parts got copied.
Besides that, I have the feeling that we should reserve some time to at least click a couple of times through staging and test different scenarios. I'm especially afraid of the Regex changes 😄

**Review by Jonasdoubleyou — APPROVED** [review]
let's go!

**Pull request merged**